doc/api/vm.md

```diff
@@ -1616,6 +1614,156 @@ inside a `vm.Context`, functions passed to them will be added to global queues,
 which are shared by all contexts. Therefore, callbacks passed to those functions
 are not controllable through the timeout either.
 
+## Support of dynamic `import()` in compilation APIs
+
+The following APIs support an `importModuleDynamically` option to enable dynamic
+`import()` in code compiled by the vm module.
+
+* `new vm.Script`
+* `vm.compileFunction()`
+* `new vm.SourceTextModule`
+* `vm.runInThisContext()`
+* `vm.runInContext()`
+* `vm.runInNewContext()`
+* `vm.createContext()`
+
+This option is still part of the experimental modules API. We do not recommend
+using it in a production environment.
+
+### When the `importModuleDynamically` option is not specified or undefined
+
+If this option is not specified, or if it's undefined, code containing
+`import()` can still be compiled by the vm APIs, but when the compiled code is
+executed and it actually calls `import()`, the result will reject with
+[`ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING`][].
+
+### When `importModuleDynamically` is `vm.constants.USE_MAIN_CONTEXT_DEFAULT_LOADER`
+
+This option is currently not supported for `vm.SourceTextModule`.
+
+With this option, when an `import()` is initiated in the compiled code, Node.js
+would use the default ESM loader from the main context to load the requested
+module and return it to the code being executed.
+
+```mjs
+import { Script, constants } from 'node:vm';
+import { resolve } from 'node:path';
+import { writeFileSync } from 'node:fs';
+
+const data = JSON.stringify({ hello: 'world' });
+writeFileSync(resolve(import.meta.dirname, 'foo.json'), data);
+
+const script = new Script('import("./foo.json", { with: { type: "json" } })', {
+  filename: resolve(import.meta.dirname, 'test-with-default.js'),
+  importModuleDynamically: constants.USE_MAIN_CONTEXT_DEFAULT_LOADER,
+});
+const result = await script.runInThisContext();
+//  { default: { hello: 'world' } }
+console.log(result);
+```
+
+```cjs
+const { Script, constants } = require('node:vm');
+const { resolve } = require('node:path');
+const { writeFileSync } = require('node:fs');
+
+const data = JSON.stringify({ hello: 'world' });
+writeFileSync(resolve(__dirname, 'foo.json'), data);
+
+const script = new Script('import("./foo.json", { with: { type: "json" } })', {
+  filename: resolve(__dirname, 'test-with-default.js'),
+  importModuleDynamically: constants.USE_MAIN_CONTEXT_DEFAULT_LOADER,
+});
+//  { default: { hello: 'world' } }
+script.runInThisContext().then(console.log);
+```
+
+There are a few caveats with loading modules using the default loader in the
+main context:
+
+1. The module being resolved would be relative to the `filename` option passed
+   to `vm.Script` or `vm.compileFunction()`. The resolution can work with a
+   `filename` that's either an absolute path or a URL string.  If `filename` is
+   a string that's neither an absolute path or a URL string, or if it's
```

Comment:
URL string, or `URL` instance? Generally we don’t allow a path to be _either_ a path string or an URL string, because of ambiguity. https://github.com/nodejs/node/issues/48994

Reply:
Only strings are allowed as resource names, passing URL instances result in an error

Reply:
Also note that this is not necessarily a path, it can be a `http://` URL. As long as `import('file:///path/to/foo')` is allowed by the default loader, this is allowed. Whenever that's becomes disallowed by the default loader, the default loader can throw, and this documentation can be updated.